File: workbench_launcher.py
"""
Workbench launcher module
"""
from browser_manager import BrowserManager
from logger_setup import setup_logger

logger = setup_logger(__name__)


class WorkbenchLauncher:
    """Handles launching the workbench from project"""
    
    def __init__(self, browser_manager: BrowserManager):
        self.browser = browser_manager
    
    async def launch_workbench(self):
        """
        Click three-dot menu and launch workbench
        """
        try:
            logger.info("Launching workbench...")
            
            # Try to find and click three-dot menu button
            menu_selectors = [
                "button[aria-label*='More']",
                "button[aria-label*='menu']",
                "button:has-text('⋮')",
                "button.menu-button",
                "button[data-testid*='menu']"
            ]
            
            menu_clicked = False
            for sel in menu_selectors:
                try:
                    await self.browser.click(sel, timeout=10000)
                    menu_clicked = True
                    logger.info(f"Clicked three-dot menu ({sel})")
                    break
                except Exception:
                    continue
            
            if not menu_clicked:
                logger.warning("Could not find three-dot menu, trying direct workbench button")
            
            # Wait a bit for menu to appear
            await self.browser.page.wait_for_timeout(500)
            
            # Click "Launch Workbench" option
            workbench_selectors = [
                "button:has-text('Launch Workbench')",
                "a:has-text('Launch Workbench')",
                "button:has-text('launch Workbench')",
                "div:has-text('Launch Workbench')",
            ]
            
            workbench_clicked = False
            for sel in workbench_selectors:
                try:
                    await self.browser.click(sel, timeout=10000)
                    workbench_clicked = True
                    logger.info(f"Clicked Launch Workbench ({sel})")
                    break
                except Exception:
                    continue
            
            if not workbench_clicked:
                raise Exception("Could not find or click 'Launch Workbench' option")
            
            # Wait for workbench to load
            await self.browser.wait_for_navigation()
            logger.info("Workbench launched successfully")
            
        except Exception as e:
            logger.error(f"Failed to launch workbench: {e}")
            raise

File: workbench_launcher.ts
// Workbench Launcher - Launches workbench
import { BrowserManager } from './browser_manager';

export class WorkbenchLauncher {
    constructor(private browser: BrowserManager) {}

    async launch(): Promise<void> {
        console.log('🚀 Launching workbench...');
        const startTime = Date.now();

        try {
            const page = this.browser.getPage();
            
            // Check if already on workbench
            const title = await page.title();
            const url = page.url();
            if (title.includes('Workbench') || url.includes('workbench')) {
                console.log('  ℹ Already on workbench');
                const duration = Date.now() - startTime;
                console.log(`✓ Already on workbench (${duration}ms)`);
                return;
            }

            // Try to find three-dot menu button (MoreVertIcon)
            console.log('  → Looking for MoreVertIcon menu button...');
            const menuFound = await this.browser.click(
                'button:has(svg[data-testid="MoreVertIcon"])||button svg[data-testid="MoreVertIcon"]',
                1000
            );

            if (menuFound) {
                // Wait for menu dropdown to appear
                await this.browser.waitForTimeout(300);
                console.log('  ℹ Menu opened, clicking Launch Workbench option');
                await this.browser.waitForTimeout(500);
                
                // Try to click workbench option - it's in a <li><button><span> structure
                const workbenchClicked = await this.browser.click(
                    'li button span:has-text("Launch Workbench")||li button:has-text("Launch Workbench")||span:has-text("Launch Workbench")',
                    1000
                );

                if (workbenchClicked) {
                    console.log('  ℹ Workbench option clicked, pausing for visibility');
                    await this.browser.waitForTimeout(1000);
                    await this.browser.waitForNavigation(5000);
                    await this.browser.waitForTimeout(300);
                    const duration = Date.now() - startTime;
                    console.log(`✓ Workbench launched (${duration}ms)`);
                    return;
                }
            }

            // If menu not found or click failed, try direct click on any "Launch" button
            console.log('  → Trying direct Launch button...');
            const directLaunch = await this.browser.click(
                'button:has-text("Launch")||a:has-text("Launch")',
                1500
            );

            if (directLaunch) {
                await this.browser.waitForNavigation(5000);
                await this.browser.waitForTimeout(300);
                const duration = Date.now() - startTime;
                console.log(`✓ Workbench launched (${duration}ms)`);
                return;
            }

            // If still not found, just continue - workbench might already be open
            console.log('  ⚠ Could not find workbench launch button, continuing...');
            const duration = Date.now() - startTime;
            console.log(`⚠ Workbench launch skipped (${duration}ms)`);

        } catch (error) {
            console.error(`✗ Workbench launch error: ${error}`);
            // Don't throw - continue anyway
        }
    }
}
